### PR TITLE
feature: v2

### PR DESCRIPTION
**Pull request by marlonmarcello**
# Description

I was about to start work on the Modal component for the component library and noticed that this could use some love so one thing led to another and I ended up rewriting it and modernizing the setup as well.

If this works well, our other libraries can follow suit and start shipping types and CSS modules.

### Changes

The old design of the modal was honestly great but it had an achilles heel, it used an internal context to handle the transition and syncing that with state which caused:
1. Actions to be tied to a context, adding custom components to close the modal for example was a bit annoying
2. Animation and state were too tightly coupled and not exposed for further customization

To solve these I decoupled things and inverted the control from the component to the hook
1. Now the modal component only cares about rendering
2. The hook takes care of the logic and expose the variables in the shape that the modal expects so the developer don't have to worry
3. It opened the door to use custom component for things like closing the modal and backdrop, so much so that we don't need to ship a close button anymore and we could honestly get away with shipping a backdrop component, it's just a positioned absolute div with a background color, too but decided to keep for convenience as that is something very common that almost never changes
4. It's much easier to reason about animation now as the duration is directly exposed
5. No need for context anymore, the hook takes care of it
6. The logic for dealing with slugs is simpler, it's automatic, you don't need to set it

This change also makes issue #2 irrelevant. Customizing the component is just a matter os styling with css and the only trick is the `transitionDuration` which is I explained in the README updates.

### New things

Modernized the library with:

1. CSS modules
2. Typescript

## Screenshots

https://github.com/wethegit/react-modal/assets/1956448/d56a884d-f157-494f-b35e-0bad1a045acd

### Simple
No custom transition, no focus management

```jsx
function MyModal() {
  const { isOpen, state, toggle } = useModal()

  return (
    <>
      <button onClick={toggle}>
        Open modal
      </button>

      {isOpen && (
        <Modal state={state}>
          <ModalBackdrop onClick={toggle} />
          <ModalContent>
            <button onClick={toggle}>
              Close
            </button>
            <p>Voluptate Lorem ut minim excepteur sit fugiat anim magna aliquip.</p>
          </ModalContent>
        </Modal>
      )}
    </>
  )
}
```

### Complete

With custom transition, focus management and hash trigger

```jsx
function MyModal() {
  const TRANSITION = 800
  const triggerButton = useRef(null)
  const { isOpen, state, toggle } = useModal({
    triggerRef: triggerButton,
    transitionDuration: TRANSITION,
    // modal will automatically open close when hash changes
    slug: "modal-with-slug",
  })

  // Custom transition must be set on the <Modal> component, how you do it, inline styles or as a className, is up to you
  const stylesVars = {
    "--transition-duration": `${TRANSITION}ms`,
  }

  return (
    <>
      {/* `triggerRef` allows the focus to shift to whatever triggered the modal, on close. */}
      <button ref={triggerButton} onClick={toggle}>
        Open the modal window!
      </button>

      {isOpen && (
        <Modal state={state} style={stylesVars}>
          <ModalBackdrop onClick={toggle} />
          <ModalContent
            className={classnames([
              // up to you how you want to handle the animation
              // you can also add separate classes for each state to customize the animation even further
              // you can also simply customize the parent <Modal> component
              // etc...
              'my-modal',
              (state === ModalStates.OPENING || state === ModalStates.OPEN) && 'my-modal-open',
            ])}
          >
            <button onClick={toggle}>
              Close
            </button>
            <p>Voluptate Lorem ut minim excepteur sit fugiat anim magna aliquip.</p>
          </ModalContent>
        </Modal>
      )}
    </>
  )
}
```

### CODE REVIEW

**Comment by marlonmarcello** [comment]
I am thinking of changing `slug` to `hash`.

File: src/lib/hooks/use-modal.ts [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+  triggerRef?: React.MutableRefObject<HTMLElement | null>
+  /**
+   * The delay (in ms) after which the modal will be closed.
+   * @defaultValue 300
+   */
+  transitionDuration?: number
+  /**
+   * If set, the modal will be opened/closed by updating the route hash.
+   */
+  slug?: string
```
Comment (andrewrubin) [inline src/lib/hooks/use-modal.ts:25]: I'm happy to change this to `hash`, as you suggested the other day, @marlonmarcello, if you like.

**Review by andrewrubin — CHANGES_REQUESTED** [review]
Great work @marlonmarcello , I dig the new setup. A few ideas and tweaks from me, below. One other thing which I cannot properly comment on — we should enforce `stylelint/order` on these CSS modules

Thanks, lmk if you want to chat about any of this!

File: src/lib/components/modal-backdrop/modal-backdrop.module.scss [hunk 2]
```diff
@@ -0,0 +1,5 @@
+.modalBackdrop {
+  background-color: rgb(0 0 0 / 80%);
```
[the remark this answers is not in the document]
Comment (marlonmarcello) [reply src/lib/components/modal-backdrop/modal-backdrop.module.scss:2]: Don't really see what's the benefit @andrewrubin 
Someone changing this component would have to:
```
.customBackdrop {
  background-color: red;
}

/* OR */

.customBackdrop {
  --backdrop-color: red
}
```

It's pretty much the exact same thing and the user don't have to know about the CSS Variable.

File: src/lib/components/modal-inner/modal-inner.module.scss [hunk 3]
```diff
@@ -0,0 +1,17 @@
+.modalInner {
+  inset: 0;
+  position: fixed;
```
[the remark this answers is not in the document]
Comment (marlonmarcello) [reply src/lib/components/modal-inner/modal-inner.module.scss:3]: Awesome idea!

File: src/lib/hooks/use-modal.ts [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    }
+  }, [handleClose, handleOpen, slug])
+
+  // Hook up the escape key
+  useEffect(() => {
+    const onKeyUp = (e: KeyboardEvent) => {
+      if (e.code === "Escape") handleClose()
+    }
+
+    window.addEventListener("keyup", onKeyUp)
```
[the remark this answers is not in the document]
Comment (marlonmarcello) [reply src/lib/hooks/use-modal.ts:137]: @andrewrubin it calls `handleClose` which takes care of that 🦄

File: src/main.module.css
[the remark this answers is not in the document]
Comment (marlonmarcello) [reply src/main.module.css:1]: Thank u!

File: src/lib/hooks/use-modal.ts [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+export interface UseModalOptions {
+  /**
+   * The ref of the element that should receive focus when the modal closes.
+   */
+  triggerRef?: React.MutableRefObject<HTMLElement | null>
+  /**
+   * The delay (in ms) after which the modal will be closed.
+   * @defaultValue 300
+   */
+  transitionDuration?: number
```
[the remark this answers is not in the document]
Comment (marlonmarcello) [reply src/lib/hooks/use-modal.ts:21]: That's a great shout

File: src/lib/hooks/use-modal.ts (hunk 1 above)
Comment (marlonmarcello) [reply src/lib/hooks/use-modal.ts:25]: Think I will then!

File: src/lib/hooks/use-modal.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (marlonmarcello) [reply src/lib/hooks/use-modal.ts:21]: Andrew, I think I am going to leave it out for now and create a separate ticket for that. That way we add it if we actually see a need for it. We did away without this far so it seems like that will be extra complexity without any usage.

[the remark this answers is not in the document]
Comment (andrewrubin) [reply src/lib/hooks/use-modal.ts:21]: I think the current solution we use is the `onOpenStart` or `onCloseStart` function, if I recall. This allows us to have finer control over when to resolve the modal as being fully closed or open, if we want to.

Maybe exposing the MODAL_STATES here solves that though?

[the remark this answers is not in the document]
Comment (marlonmarcello) [reply src/lib/hooks/use-modal.ts:21]: I did expose the `state` on the return. But still, do we use those?

[the remark this answers is not in the document]
Comment (andrewrubin) [reply src/lib/hooks/use-modal.ts:21]: Nah I don't think we really do :) But we totally could… tbh I'm just thinking out loud at this point. 

Interesting idea though, for example, maybe sometime down the road we decide we want an option to specify exactly when these things finished opening and closing, _outside_ the realm of CSS or setTimeouts. Silly example:

```jsx
async function handleOpen () {
  await loadTwentyGIFsFromGiphy()
  dispatchState(MODAL_STATES.OPENED)
}

<Modal onOpen={handleOpen} />
```

Could be a fun problem to solve. But not necessary right now I don't think.

File: src/lib/hooks/use-modal.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (andrewrubin) [reply src/lib/hooks/use-modal.ts:137]: Chatted over Slack –  this still needs to remove the hash on Esc, so we discussed moving all hash-related logic into `handleOpen` and `handleClose`

File: src/lib/hooks/use-modal.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (marlonmarcello) [reply src/lib/hooks/use-modal.ts:21]: We should look into how react transition group and framer do this, having variables for each in/out/in-between sounds so inneficient

Maybe we do that now, with this new major @andrewrubin

File: src/lib/hooks/use-modal.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (marlonmarcello) [reply src/lib/hooks/use-modal.ts:137]: Tomorrow I am going to see if I can reduce the amount of useEffects and callback and update this

File: src/lib/hooks/use-modal.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (andrewrubin) [reply src/lib/hooks/use-modal.ts:21]: As far as my experience with RTG goes, I know they use four different states, sort of similar: entering, entered, exiting, and exited. Those are passed down through a state argument to `children(state)` if I recall correctly. 

As for framer, it's clean in that they abstract a lot away, and allow you to just write your "shown" and "hidden" states and define the transitions between them... but I don't like the api that much, it feels like, at least **for our purposes**, too much of an abstraction, when we could easily use simple CSS for things.

[the remark this answers is not in the document]
Comment (marlonmarcello) [reply src/lib/hooks/use-modal.ts]: Kay. I removed this entirely. The modal shouldn't have to take care of animation like this. I am moving this responsibility to a [hook](https://github.com/wethegit/component-library/pull/63).

**Comment by marlonmarcello** [comment]
All done now @andrewrubin

**Comment by marlonmarcello** [comment]
Converting it to a Draft while we wait for [hooks v2](https://github.com/wethegit/react-hooks/pull/6)

**Review by andrewrubin — APPROVED** [review]

**Pull request merged**